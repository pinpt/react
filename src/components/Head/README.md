### A head element to handle the required includes and metadata tags for SEO, etc

[Full Docs](react.preview.pinpoint.com/?path=/docs/components-head) | [Github](https://github.com/pinpt/react/tree/master/src/components/Head)

#### Theming

This component has no styles associated with it.
