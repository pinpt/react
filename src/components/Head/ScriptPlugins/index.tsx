import React from 'react';
import GoogleTagManagerPlugin from './gtm';
import TypekitPlugin from './typekit';

const pluginRegistry: Record<string, (opts: any) => React.ReactFragment> = {
	gtm: (opts) => <GoogleTagManagerPlugin options={opts} key="gtm" />,
	typekit: (opts) => <TypekitPlugin options={opts} key="typekit" />,
};

const ScriptPlugins = ({ plugins }: { plugins: Record<string, any> }) => {
	return (
		<>
			{Object.keys(plugins)
				.filter((pluginKey) => {
					if (pluginRegistry[pluginKey]) {
						return true;
					}
					console.error(`Plugin ${pluginKey} not found`);
					return false;
				})
				.map((key) => pluginRegistry[key](plugins[key]))}
		</>
	);
};

export default ScriptPlugins;
