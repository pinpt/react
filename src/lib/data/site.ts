import type { Site } from '../types/site';
import { executeAPI } from '../fetch';

export const fetchSite = async (siteId: string): Promise<Site> => {
	const { site } = await executeAPI(`/site-api/v1/site/${siteId}`);
	return site;
};

type AnalyticsResult = Record<string, { claps: number; pageviews: number }>;

export const fetchAnalytics = async (siteId: string): Promise<AnalyticsResult> => {
	const { data } = await executeAPI(`/site-api/v1/site/${siteId}/analytics`);
	return data;
};
