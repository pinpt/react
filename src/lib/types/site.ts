export interface ISite {
	id: string;
	logoUrl?: string;
	mailingAddress?: string;
	url: string; // the canonical url for the site
	slug: string;
	name: string;
	private: boolean; // true if private and should not be indexable by search engines
	theme?: {
		title?: string;
		social?: Record<string, string>;
		description?: string;
		logoLink?: string;
		copyright?: string;
		eulaLink?: string;
		privacyLink?: string;
		tosLink?: string;
		homepage?: {
			tags?: string[];
		};
	};
	scriptPlugins?: Record<string, any>;
	basePath: string;
}
