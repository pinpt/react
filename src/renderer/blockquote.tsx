import { registerNode, recurseIntoChildren, NodeProps } from './register';

const BlockQuote = ({ node }: NodeProps) => {
	return <blockquote>{recurseIntoChildren(node)}</blockquote>;
};

registerNode('blockquote', (node) => <BlockQuote node={node} />);
