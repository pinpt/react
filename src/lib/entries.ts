import { Entry } from './types';

export const splitEntries = (entries: Entry[] = [], latestCount = 1) => {
	const latest = entries.slice(0, latestCount);
	const recent = entries.slice(latestCount, entries.length);

	return { latest, recent };
}