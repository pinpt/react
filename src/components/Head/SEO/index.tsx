import type { ISite, IContent } from '../../../lib/types';
import React from 'react';
import { extractImageMetadataFromFileID } from '../../../lib/file_metadata';
import { getTwitterProfileFromURL } from '../../../lib/string';

export interface ISEOProps {
	site: ISite;
	content?: IContent;
}

const SiteSEO = ({ site }: { site: ISite }) => {
	const title = site.theme?.description ? `${site.theme.description} - ${site.name}` : site.name;
	const description = site.theme?.description ?? site.name;
	return (
		<>
			<title>{title}</title>
			<meta name="type" property="og:type" content="website" />
			<meta name="title" property="og:title" content={title} />
			<meta name="twitter:title" content={title} />
			<meta name="description" property="og:description" content={description} />
			<meta name="twitter:description" content={description} />
		</>
	);
};

const formatAuthors = (content: IContent): string => {
	const authors =
		content.authors?.filter((author) => !!author).map((author) => `${author.firstName} ${author.lastName}`) ?? [];
	return authors.join(', ');
};

const getCoverImage = (content: IContent): string => {
	if (content.coverMedia?.placeholderImage) {
		const { size } = extractImageMetadataFromFileID(content.coverMedia.placeholderImage);
		if (size?.width !== 1200 && size?.height !== 628) {
			const u = new URL(content.coverMedia.placeholderImage);
			u.searchParams.set('rw', '1200');
			u.searchParams.set('rh', '628');
			return u.toString();
		}
		return content.coverMedia.placeholderImage;
	}
	return '';
};

const ContentSEO = ({ site, content }: { site: ISite; content: IContent }) => {
	const title = content.title;
	const description = content.headline;
	const author = formatAuthors(content);
	const date = new Date(content.publishedAt).toISOString();
	const coverImage = getCoverImage(content);
	return (
		<>
			<title>
				{title} - {site.name}
			</title>
			{coverImage && (
				<>
					<meta name="image" property="og:image" content={coverImage} />
					<meta name="twitter:image" content={coverImage} />
					<meta name="twitter:image:alt" content={title} />
				</>
			)}
			<meta name="type" property="og:type" content="article" />
			<meta property="author" content={author} />
			<meta name="author" property="og:author" content={author} />
			<meta property="article:published_time" content={date} />
			<meta name="title" property="og:title" content={title} />
			<meta name="twitter:title" content={title} />
			<meta name="description" property="og:description" content={description} />
			<meta name="twitter:description" content={description} />
			<link
				rel="alternate"
				type="application/rss+xml"
				title={`RSS Feed for ${site.theme?.title ?? site.name}`}
				href={`${site.url}/rss`}
			/>
		</>
	);
};

const SEO = (props: ISEOProps) => {
	const { site, content } = props;
	const contentSEO = content ? <ContentSEO site={site} content={content} /> : null;
	const siteSEO = !content ? <SiteSEO site={site} /> : null;
	const url = content ? content.url : site.url;
	const twitterHandle = getTwitterProfileFromURL(site.theme?.social?.twitter);
	return (
		<>
			{contentSEO}
			{siteSEO}
			<link rel="icon" href={site.logoUrl} />
			<link rel="shortcut icon" href={site.logoUrl} />
			<link rel="alternate" type="application/rss+xml" title={`${site.name} RSS Feed`} href="/rss" />
			<meta property="og:site_name" content={site.name} />
			<meta name="robots" content="index, follow" />
			<meta name="twitter:card" content="summary_large_image" />
			<meta property="og:url" content={url} />
			<meta property="twitter:url" content={url} />
			<link rel="canonical" href={url} />
			{twitterHandle && (
				<>
					<meta name="twitter:creator" content={twitterHandle} />
					<meta name="twitter:site" content={twitterHandle} />
				</>
			)}
		</>
	);
};

export default SEO;
