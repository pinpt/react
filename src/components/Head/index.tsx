import React from 'react';
import { getSiteAnalyticsURL } from '../../lib/router';
import { IContent } from '../../lib/types/content';
import { ISite } from '../../lib/types/site';
import SEO from './SEO';
import ScriptPlugins from './ScriptPlugins';

export interface IHeadProps {
	site: ISite;
	content?: IContent;
	children?: React.ReactChild;
}

const Head = (props: IHeadProps) => {
	const { site, content, children } = props;
	return (
		<>
			<meta charSet="utf-8" />
			<link rel="preconnect" href="https://site-cdn.pinpoint.com" />
			<link rel="preconnect" href="https://file.pinpoint.com" />
			<meta httpEquiv="x-ua-compatible" content="ie=edge" />
			<meta name="viewport" content="width=device-width" />
			<script
				src={getSiteAnalyticsURL(site)}
				data-site-id={site.id}
				data-id={content?.id}
				data-base-path={site.basePath}
				data-use-react
			/>
			{site.scriptPlugins ? <ScriptPlugins plugins={site.scriptPlugins} /> : null}
			<meta name="generator" content="pinpoint.com" />
			<SEO site={site} content={content} />
			{children}
		</>
	);
};

export default Head;
