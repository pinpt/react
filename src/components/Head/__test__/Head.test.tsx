import renderer from 'react-test-renderer';
import Head from '../';
import type { IContent } from '../../../lib/types';
import entries from '../../Prebuilt/__data__/testEntries.json';
import site from '../../Prebuilt/__data__/testSite.json';

test('Test default render for Site', () => {
	const component = renderer.create(<Head site={site} />);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Test default render for Content', () => {
	const component = renderer.create(<Head site={site} content={entries[0] as IContent} />);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

test('Test setting basePath', () => {
	const component = renderer.create(<Head site={{ ...site, basePath: '/blog' }} content={entries[0] as IContent} />);
	const tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});
