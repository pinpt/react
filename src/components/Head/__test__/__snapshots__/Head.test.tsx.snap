// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test default render for Content 1`] = `
Array [
  <meta
    charSet="utf-8"
  />,
  <meta
    content="ie=edge"
    httpEquiv="x-ua-compatible"
  />,
  <meta
    content="width=device-width"
    name="viewport"
  />,
  <script
    async={true}
    data-id="oroH8YQr4vufOiVTRND4"
    data-site-id="PirxVTE94u3YmGNOySRY"
    defer={true}
    src="/a.js"
  />,
  <meta
    content="pinpoint.com"
    name="generator"
  />,
  <title>
    Linked Content Blocks
     - 
    Pinpoint
  </title>,
  <meta
    content="https://file.pinpoint.com/dee04c201d3ac96f3314041b5382cc43;ULH_.DD%2500%25Ms%3Bxbxbj%5D00%25M~qRjjKV%5BRjax;1600x900.png?rw=1200&rh=628"
    name="image"
    property="og:image"
  />,
  <meta
    content="https://file.pinpoint.com/dee04c201d3ac96f3314041b5382cc43;ULH_.DD%2500%25Ms%3Bxbxbj%5D00%25M~qRjjKV%5BRjax;1600x900.png?rw=1200&rh=628"
    name="twitter:image"
  />,
  <meta
    content="Linked Content Blocks"
    name="twitter:image:alt"
  />,
  <meta
    content="article"
    name="type"
    property="og:type"
  />,
  <meta
    content="Chris Bowley"
    property="author"
  />,
  <meta
    content="Chris Bowley"
    name="author"
    property="og:author"
  />,
  <meta
    content="2021-08-02T14:16:50.147Z"
    property="article:published_time"
  />,
  <meta
    content="Linked Content Blocks"
    name="title"
    property="og:title"
  />,
  <meta
    content="Linked Content Blocks"
    name="twitter:title"
  />,
  <meta
    content="Linked content blocks allow you to populate your content from the recently completed work. Annotations show you which issues and pull requests are linked."
    name="description"
    property="og:description"
  />,
  <meta
    content="Linked content blocks allow you to populate your content from the recently completed work. Annotations show you which issues and pull requests are linked."
    name="twitter:description"
  />,
  <link
    href="https://file.pinpoint.com/1fcde4196a4c70a8a86f0ce4af53f2a5;U75%3B-sj_NAoPtDojoij_RdfPRdaxoPfPavaw;320x320.png"
    rel="icon"
  />,
  <link
    href="https://file.pinpoint.com/1fcde4196a4c70a8a86f0ce4af53f2a5;U75%3B-sj_NAoPtDojoij_RdfPRdaxoPfPavaw;320x320.png"
    rel="shortcut icon"
  />,
  <link
    href="/rss"
    rel="alternate"
    title="Pinpoint RSS Feed"
    type="application/rss+xml"
  />,
  <meta
    content="Pinpoint"
    property="og:site_name"
  />,
  <meta
    content="index, follow"
    name="robots"
  />,
  <meta
    content="summary_large_image"
    name="twitter:card"
  />,
  <meta
    content="https://changelog.pinpoint.com/entry/atBB4DPW5UrCgugTPw6M/Linked-Content-Blocks"
    property="og:url"
  />,
  <meta
    content="https://changelog.pinpoint.com/entry/atBB4DPW5UrCgugTPw6M/Linked-Content-Blocks"
    property="twitter:url"
  />,
  <link
    href="https://changelog.pinpoint.com/entry/atBB4DPW5UrCgugTPw6M/Linked-Content-Blocks"
    rel="canonical"
  />,
  <meta
    content="@pinpoint_sw"
    name="twitter:creator"
  />,
  <meta
    content="@pinpoint_sw"
    name="twitter:site"
  />,
]
`;

exports[`Test default render for Site 1`] = `
Array [
  <meta
    charSet="utf-8"
  />,
  <meta
    content="ie=edge"
    httpEquiv="x-ua-compatible"
  />,
  <meta
    content="width=device-width"
    name="viewport"
  />,
  <script
    async={true}
    data-site-id="PirxVTE94u3YmGNOySRY"
    defer={true}
    src="/a.js"
  />,
  <meta
    content="pinpoint.com"
    name="generator"
  />,
  <title>
    New updates and improvements to Pinpoint - Pinpoint
  </title>,
  <meta
    content="website"
    name="type"
    property="og:type"
  />,
  <meta
    content="New updates and improvements to Pinpoint - Pinpoint"
    name="title"
    property="og:title"
  />,
  <meta
    content="New updates and improvements to Pinpoint - Pinpoint"
    name="twitter:title"
  />,
  <meta
    content="New updates and improvements to Pinpoint"
    name="description"
    property="og:description"
  />,
  <meta
    content="New updates and improvements to Pinpoint"
    name="twitter:description"
  />,
  <link
    href="https://file.pinpoint.com/1fcde4196a4c70a8a86f0ce4af53f2a5;U75%3B-sj_NAoPtDojoij_RdfPRdaxoPfPavaw;320x320.png"
    rel="icon"
  />,
  <link
    href="https://file.pinpoint.com/1fcde4196a4c70a8a86f0ce4af53f2a5;U75%3B-sj_NAoPtDojoij_RdfPRdaxoPfPavaw;320x320.png"
    rel="shortcut icon"
  />,
  <link
    href="/rss"
    rel="alternate"
    title="Pinpoint RSS Feed"
    type="application/rss+xml"
  />,
  <meta
    content="Pinpoint"
    property="og:site_name"
  />,
  <meta
    content="index, follow"
    name="robots"
  />,
  <meta
    content="summary_large_image"
    name="twitter:card"
  />,
  <meta
    content="https://changelog.pinpoint.com"
    property="og:url"
  />,
  <meta
    content="https://changelog.pinpoint.com"
    property="twitter:url"
  />,
  <link
    href="https://changelog.pinpoint.com"
    rel="canonical"
  />,
  <meta
    content="@pinpoint_sw"
    name="twitter:creator"
  />,
  <meta
    content="@pinpoint_sw"
    name="twitter:site"
  />,
]
`;
